tests/decodeoutput.cpp

```diff
@@ -738,6 +741,205 @@ DecodeOutputDmabuf::DecodeOutputDmabuf(VideoDataMemoryType memoryType)
+                                              const VARectangle *dstRect)
+{
+    VAStatus vaStatus;
+    struct wl_buffer *buffer;
+    struct wl_callback *callback;
+    struct display * const d = &m_waylandDisplay;
+    struct wl_callback_listener frame_callback_listener = {frameRedrawCallback};
+
+    if (m_redrawPending) {
+        wl_display_flush(d->display);
+        while (m_redrawPending) {
+            wl_display_dispatch(d->display);
+        }
+    }
+
+    if (!ensureWindow(dstRect->width, dstRect->height))
+        return false;
+
+    vaStatus = vaGetSurfaceBufferWl(m_vaDisplayHandle, surface, VA_FRAME_PICTURE, &buffer);
+    if (vaStatus != VA_STATUS_SUCCESS)
+        return false;
+
+    wl_surface_attach(d->surface, buffer, 0, 0);
+    wl_surface_damage(d->surface, dstRect->x,
+		dstRect->y, dstRect->width, dstRect->height);
+    wl_display_flush(d->display);
+    m_redrawPending = true;
+    callback = wl_surface_frame(d->surface);
+    wl_surface_commit(d->surface);
+    wl_callback_add_listener(callback, &frame_callback_listener,&m_redrawPending);
+    return true;
+}
+
+bool DecodeOutputWayland::output(const SharedPtr<VideoFrame>& frame)
+{
+    VARectangle srcRect, dstRect;
+    if (!setVideoSize(frame->crop.width, frame->crop.height))
+        return false;
+
+    srcRect.x = 0;
+    srcRect.y = 0;
+    srcRect.width  = frame->crop.width;
+    srcRect.height = frame->crop.height;
+
+    dstRect.x = frame->crop.x;
+    dstRect.y = frame->crop.y;
+    dstRect.width  = frame->crop.width;
+    dstRect.height = frame->crop.height;
+    return vaPutSurfaceWayland((VASurfaceID)frame->surface, &srcRect, &dstRect);
+}
+
+bool DecodeOutputWayland::setVideoSize(uint32_t width, uint32_t height)
+{
+    return ensureWindow(width, height);
+}
+
+bool DecodeOutputWayland::createWaylandDisplay()
+{
+    int major, minor;
+    SharedPtr<VADisplay> display;
+    struct display *d = &m_waylandDisplay;
+    struct wl_registry_listener registry_listener = {
+        DecodeOutputWayland::registryHandle,
+        NULL,
+    };
+
+    d->display = wl_display_connect(NULL);
+    if (!d->display) {
+        return false;
+    }
+    wl_display_set_user_data(d->display, d);
+    d->registry = wl_display_get_registry(d->display);
+    wl_registry_add_listener(d->registry, &registry_listener, d);
+    wl_display_dispatch(d->display);
+    m_vaDisplayHandle= vaGetDisplayWl(d->display);
+    VAStatus status = vaInitialize(m_vaDisplayHandle, &major, &minor);
+    if (!checkVaapiStatus(status, "vaInitialize"))
+        return false;
+    m_vaDisplay.reset(new VADisplay(m_vaDisplayHandle), VADisplayTerminator());
+    return true;
+}
+
+bool DecodeOutputWayland::init()
+{
+    return createWaylandDisplay() && DecodeOutput::init();
+}
+
+DecodeOutputWayland::DecodeOutputWayland()
+    : m_redrawPending(false)
+{
+}
+
+DecodeOutputWayland::~DecodeOutputWayland()
+{
+    struct display * const d = &m_waylandDisplay;
```

Comment:
We did not memset m_waylandDisplay..., so this will crash if something wrong.
Could we put this to m_waylandDisplay's deconstructor?